File: Editorial/0000-0100/0076_Minimum_Window_Substring/CSharp/1_Sliding_Window.cs
public class Solution {
    public string MinWindow(string s, string t) {
        if (s.Length == 0 || t.Length == 0) {
            return "";
        }

        // Dictionary which keeps a count of all the unique characters in t.
        Dictionary<char, int> dictT = new Dictionary<char, int>();
        for (int i = 0; i < t.Length; i++) {
            if (dictT.ContainsKey(t[i])) {
                dictT[t[i]]++;
            } else {
                dictT[t[i]] = 1;
            }
        }

        // Number of unique characters in t, which need to be present in the
        // desired window.
        int required = dictT.Count;
        // Left and Right pointer
        int l = 0, r = 0;
        // formed is used to keep track of how many unique characters in t
        // are present in the current window in its desired frequency.
        // e.g. if t is "AABC" then the window must have two A's, one B and one
        // C. Thus formed would be = 3 when all these conditions are met.
        int formed = 0;
        // Dictionary which keeps a count of all the unique characters in the
        // current window.
        Dictionary<char, int> windowCounts = new Dictionary<char, int>();
        // ans list of the form (window length, left, right)
        int[] ans = { -1, 0, 0 };
        while (r < s.Length) {
            // Add one character from the right to the window
            char c = s[r];
            if (windowCounts.ContainsKey(c)) {
                windowCounts[c]++;
            } else {
                windowCounts[c] = 1;
            }

            // If the frequency of the current character added equals to the
            // desired count in t then increment the formed count by 1.
            if (dictT.ContainsKey(c) && windowCounts[c] == dictT[c]) {
                formed++;
            }

            // Try and contract the window till the point where it ceases to be
            // 'desirable'.
            while (l <= r && formed == required) {
                c = s[l];
                // Save the smallest window until now.
                if (ans[0] == -1 || r - l + 1 < ans[0]) {
                    ans[0] = r - l + 1;
                    ans[1] = l;
                    ans[2] = r;
                }

                // The character at the position pointed by the
                // `Left` pointer is no longer a part of the window.
                windowCounts[c]--;
                if (dictT.ContainsKey(c) && windowCounts[c] < dictT[c]) {
                    formed--;
                }

                // Move the left pointer ahead, this would help to look for a
                // new window.
                l++;
            }

            // Keep expanding the window once we are done contracting.
            r++;
        }

        return ans[0] == -1 ? "" : s.Substring(ans[1], ans[2] - ans[1] + 1);
    }
}

File: Editorial/0101-0200/0159_Longest_Substring_with_At_Most_Two_Distinct_Characters/TypeScript/1_Sliding_Window.ts
function lengthOfLongestSubstringTwoDistinct(s: string): number {
    let n = s.length;
    if (n < 3) return n;

    // sliding window left and right pointers
    let left = 0;
    let right = 0;
    // hashmap character -> its rightmost position
    // in the sliding window
    let hashmap: { [key: string]: number } = {};

    let max_len = 2;

    while (right < n) {
        // when the slidewindow contains less than 3 characters
        hashmap[s[right]] = right++;
        // slidewindow contains 3 characters
        if (Object.keys(hashmap).length == 3) {
            // delete the leftmost character
            let del_idx = Math.min(...Object.values(hashmap));
            delete hashmap[s[del_idx]];
            // move left pointer of the slidewindow
            left = del_idx + 1;
        }

        max_len = Math.max(max_len, right - left);
    }
    return max_len;
}
